### PR TITLE
[T-242] Docker for alpine and centos binaries

### PR DESCRIPTION
**Pull request by DeMoorJasper**
This PR aims to create binaries for alpine and linux versions that use an older libc than the one we use in CI (apparently centos is a good target to build against as it's using a very old libc version... source: `prebuild`)

Using docker this PR virtualises `centos` and `alpine` using the docker images by the `prebuild` project.

Related #23 #18 
Closes T-242

### CODE REVIEW

**Comment by mischnic** [comment]
- The alpine issue: Alpine uses musl, which is a reimplementation of glibc. It seems like the fts functions (which are part of glibc) aren't installed by default. They can be installed with `apk add fts-dev` (which should install the `fts` package as well?), which should hopefully fix the error.

- The centos problem:
> Support for 64-bit operations in fts functions was added only a few years ago in glibc through this commit and there are systems that still ship older versions glibc (for e.g. opensuse) [https://github.com/mesonbuild/meson/issues/3519]
So 64-bit support for fts was only added in glibc 2.23: https://abi-laboratory.pro/?view=changelog&l=glibc&v=2.23

Not quite sure what we can/should do here

**Comment by DeMoorJasper** [comment]
@mischnic awesome, thanks for the fast review.

Knew alpine ran musl didn't know you needed to install that package, thanks appears to have fixed it.

Read about centos only supporting 32bit fts, that's what I meant with superold. Didn't even know 32-bit was still a thing. Have no clue if it's possible to build 2 versions and just switch between 32 and 64 bit fts

**Comment by mischnic** [comment]
prebuild might support fallback to 32bit, but having libraries with different architectures can cause problems on linux: if the 32bit watcher library links dynamically to glibc, then glibc needs to be installed in a 32bit variant as well. I'm not sure how this works out with centos.

**Comment by mischnic** [comment]
Another possibility: loading glibc via `dlopen`, so that we can handle a failure ourselves and fallback to the `BruteForceBackend`.

**Comment by DeMoorJasper** [comment]
Isn't it used for the BruteForceBackend?

**Comment by mischnic** [comment]
What do you mean?

**Comment by DeMoorJasper** [comment]
@mischnic I mean it's used in `BruteForceBackend::readTree` on unix based systems (mac and linux)

So we can't really fallback to `BruteForceBackend` as this is already `BruteForceBackend`

**Comment by mischnic** [comment]
Ah, you're right

**Comment by DeMoorJasper** [comment]
Wrote a custom directory traverser, unfortunately not all tests are passing :(

Not entirely sure why though, I'm probably not familiar enough with the watcher codebase...

Performance wise it's not that much slower that we should really worry especially as this is a fallback anyway afaik...

| implementation | runtime | detected files |
| --- | --- | --- |
| fts | `590ms` | 55208 |
| custom | `720ms (+22%)` | 55207  |

File: binding.gyp [hunk 1]
```diff
@@ -35,7 +35,7 @@
             "src/watchman/WatchmanBackend.cc",
             "src/shared/BruteForceBackend.cc",
             "src/linux/InotifyBackend.cc",
-            "src/unix/fts.cc"
+            "src/unix/legacy.cc"
```
Comment (DeMoorJasper) [inline binding.gyp]: TODO: Figure out a way to use fts when using musl, as alpine includes Large file system support and builds fts just fine :)

File: src/unix/fts.cc [hunk 2]
```diff
@@ -1,5 +1,9 @@
 #include <string>
-#define __THROW // weird error on linux
+
+// weird error on linux
+#undef __THROW
```
Comment (DeMoorJasper) [inline src/unix/fts.cc]: Adding this fixes a warning that pops up on build

File: package.json [hunk 3]
```diff
@@ -9,6 +9,7 @@
   "scripts": {
     "install": "prebuild-install -r napi || node-gyp rebuild",
     "prebuild": "prebuild -t 3 -r napi --strip",
+    "prebuild:docker": "make build && make run",
```
Comment (DeMoorJasper) [inline package.json]: Figure out how to run this in CI...

File: src/unix/fts.cc (hunk 2 above)
Comment (mischnic) [reply src/unix/fts.cc]: To be sure that this doesn't cause a compile error on some systems:
```c
#ifdef __THROW
#undef __THROW
#endif
#define __THROW
```

Comment (DeMoorJasper) [reply src/unix/fts.cc]: Thanks for this :)

**Comment by DeMoorJasper** [comment]
Some inotify tests are failing on linux but not sure why, will try to get my hands on a linux vm or something to debug it a bit...

**Pull request merged**